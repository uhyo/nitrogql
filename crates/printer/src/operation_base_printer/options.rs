use nitrogql_config_file::Config;
use nitrogql_utils::clone_into;

/// Options for OperationBasePrinter.
#[derive(Debug, Clone)]
pub struct OperationBasePrinterOptions {
    /// Whether an operation should be default exported when it is the only operation in the document.
    pub default_export_for_operation: bool,
    /// Whether an operation should be named exported.
    pub named_export_for_operation: bool,
    /// Whether an operation input type should be exported.
    pub export_input_type: bool,
    /// Whether an operation result type should be exported.
    pub export_result_type: bool,
    /// Whether operation name should be capitalize
    pub capitalize_operation_names: bool,
    /// Suffix for variable of query.
    pub query_variable_suffix: String,
    /// Suffix for variable of mutation.
    pub mutation_variable_suffix: String,
    /// Suffix for variable of subscription.
    pub subscription_variable_suffix: String,
    /// Suffix for variable of fragment.
    pub fragment_variable_suffix: String,
}

impl Default for OperationBasePrinterOptions {
    fn default() -> Self {
        Self {
            default_export_for_operation: true,
            named_export_for_operation: false,
            export_input_type: false,
            export_result_type: false,
            capitalize_operation_names: true,
            query_variable_suffix: "Query".to_owned(),
            mutation_variable_suffix: "Mutation".to_owned(),
            subscription_variable_suffix: "Subscription".to_owned(),
            fragment_variable_suffix: "".to_owned(),
        }
    }
}

impl OperationBasePrinterOptions {
    /// Creates a new instance of OperationBasePrinterOptions from
    /// Config.
    pub fn from_config(config: &Config) -> Self {
        let mut result = Self {
            default_export_for_operation: config.generate.export.default_export_for_operation,
            named_export_for_operation: !config.generate.export.default_export_for_operation,
            export_input_type: config.generate.export.variables_type,
            export_result_type: config.generate.export.operation_result_type,
            ..Self::default()
        };
        clone_into(
            &config.generate.name.capitalize_operation_names,
            &mut result.capitalize_operation_names,
        );
        clone_into(
            &config.generate.name.query_variable_suffix,
            &mut result.query_variable_suffix,
        );
        clone_into(
            &config.generate.name.mutation_variable_suffix,
            &mut result.mutation_variable_suffix,
        );
        clone_into(
            &config.generate.name.subscription_variable_suffix,
            &mut result.subscription_variable_suffix,
        );
        clone_into(
            &config.generate.name.fragment_variable_suffix,
            &mut result.fragment_variable_suffix,
        );
        result
    }
}
