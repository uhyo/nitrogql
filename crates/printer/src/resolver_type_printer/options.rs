use nitrogql_config_file::Config;

#[derive(Debug)]
pub struct ResolverTypePrinterOptions {
    /// Name of root resolver type.
    pub root_resolver_type: String,
    /// Name of resolver output type.
    pub resolver_output_type: String,
    /// Source of schema type to import from.
    pub schema_source: String,
    /// Name of the root TypeScript namespace that contains schema types.
    pub schema_root_namespace: String,
}

impl Default for ResolverTypePrinterOptions {
    fn default() -> Self {
        ResolverTypePrinterOptions {
            root_resolver_type: "Resolvers".into(),
            resolver_output_type: "ResolverOutput".into(),
            schema_source: "".into(),
            schema_root_namespace: "Schema".into(),
        }
    }
}

impl ResolverTypePrinterOptions {
    pub fn from_config(_config: &Config) -> Self {
        ResolverTypePrinterOptions::default()
    }
}
