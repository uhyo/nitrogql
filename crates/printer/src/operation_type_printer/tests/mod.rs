use std::path::Path;

use insta::assert_snapshot;

use graphql_builtins::generate_builtins;
use nitrogql_ast::{set_current_file_of_pos, OperationDocumentExt};
use nitrogql_ast::{OperationDocument, TypeSystemDocument};
use nitrogql_parser::{parse_operation_document, parse_type_system_document};
use nitrogql_semantics::{
    ast_to_type_system, resolve_operation_extensions, OperationExtension, OperationResolver,
};
use nitrogql_semantics::{resolve_operation_imports, resolve_schema_extensions};
use sourcemap_writer::JustWriter;

use crate::operation_base_printer::options::OperationBasePrinterOptions;
use crate::print_types_for_operation_document;
use crate::OperationTypePrinterOptions;

fn type_system() -> TypeSystemDocument<'static> {
    let mut doc = parse_type_system_document(
        "
            type User implements HasID {
                id: ID!
                name: String!
                type: UserType!
                age: Int
                posts: [HasID!]!
            }
            interface HasID {
                id: ID!
            }
            enum UserType {
                NormalUser
                PremiumUser
            }
            type Bot implements HasID {
                id: ID!
            }

            type Post implements HasID {
                id: ID!
                title: String!
                tags: [String!]
                body: String!
            }

            type Tweet implements HasId {
                id: ID!
                body: String!
            }

            input UserSearchQuery {
                age: Int
                name: String
                keywords: [String!]
            }

            type Query {
                me: User!
            }
            ",
    )
    .unwrap();
    doc.extend(generate_builtins());
    let doc = resolve_schema_extensions(doc).unwrap();
    doc
}

#[test]
fn basic_type_printing() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn export_input_and_result_type() {
    let doc = parse_operation_document(
        "
        query sampleQuery {
            me {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let (doc, _) = resolve_operation_extensions(doc).unwrap();
    let printed = print_document(
        &doc,
        OperationTypePrinterOptions {
            base_options: OperationBasePrinterOptions {
                export_input_type: true,
                export_result_type: true,
                ..Default::default()
            },
            ..Default::default()
        },
    );
    assert_snapshot!(printed);
}

#[test]
fn fragment_spread() {
    let doc = parse_operation_document(
        "
        query test{
            me {
                id name type age
                posts {
                    id
                    ...F
                }
            }
        }
        fragment F on HasID {
            id
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_spread_cond() {
    let doc = parse_operation_document(
        "
        query fooBar123 {
            me {
                id name type age
                posts {
                    id
                    ...F
                    ...P
                }
            }
        }
        fragment F on HasID {
            id
        }
        fragment P on Post {
            title
            body
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_inline_spread() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name type age
                posts {
                    ... {
                        id
                    }
                }
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_merging() {
    let doc = parse_operation_document(
        "
        query {
            me {
                ...F
                ...P
            }
        }
        fragment F on User {
            id
        }
        fragment P on User {
            name
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn fragment_nested_merging() {
    let doc = parse_operation_document(
        "
        query {
            ...F
            ...P
        }
        fragment F on Query {
            me {
                id
            }
        }
        fragment P on Query {
            me {
                name
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn query_variables() {
    let doc = parse_operation_document(
        "
        query testQuery($foo: Int!, $bar: String) {
            me(foo: $foo, bar: $bar) {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let printed = print_document_default(&doc);
    assert_snapshot!(printed);
}

#[test]
fn variable_disallow_undefined() {
    let doc = parse_operation_document(
        "
        query testQuery($foo: Int!, $bar: String) {
            me(foo: $foo, bar: $bar) {
                id name type age
            }
        }
        ",
    )
    .unwrap();
    let (doc, _) = resolve_operation_extensions(doc).unwrap();
    let printed = print_document(
        &doc,
        OperationTypePrinterOptions {
            base_options: Default::default(),
            allow_undefined_as_optional_input: false,
            ..Default::default()
        },
    );
    assert_snapshot!(printed);
}

#[test]
fn print_values() {
    let doc = parse_operation_document(
        "
        query {
            me {
                id name
            }
        }
        ",
    )
    .unwrap();
    let (doc, _) = resolve_operation_extensions(doc).unwrap();
    let mut result = String::new();
    let mut writer = JustWriter::new(&mut result);
    let schema = type_system();
    let schema = ast_to_type_system(&schema);
    let options = OperationTypePrinterOptions {
        print_values: true,
        ..Default::default()
    };
    print_types_for_operation_document(options, &schema, &doc, &mut writer);
    assert_snapshot!(result);
}

mod skip_include {
    use super::*;

    #[test]
    fn skipped_field_is_not_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @skip(if: true)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn not_skipped_field_is_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @skip(if: false)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn included_field_is_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @include(if: true)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn not_included_field_is_not_included() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    age @include(if: false)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_variable() {
        let doc = parse_operation_document(
            "
            query($skip: Boolean!) {
                me {
                    id name type
                    age @skip(if: $skip)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn include_variable() {
        let doc = parse_operation_document(
            "
            query($include: Boolean!) {
                me {
                    id name type
                    age @include(if: $include)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_and_include() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id @skip(if: false) @include(if: false)
                    name @skip(if: true) @include(if: false)
                    type @skip(if: false) @include(if: true)
                    age @skip(if: true) @include(if: true)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn emit_only_possible_cases() {
        let doc = parse_operation_document(
            "
            query($flag: Boolean!) {
                me {
                    id @skip(if: $flag)
                    name @include(if: $flag)
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_fragment_spread() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    ...F @skip(if: true)
                }
            }
            fragment F on User {
                age
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn include_fragment_spread() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    ...F @include(if: true)
                }
            }
            fragment F on User {
                age
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_inline_fragment() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    ... @skip(if: true) {
                        age
                    }
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn include_inline_fragment() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    ... @include(if: true) {
                        age
                    }
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_and_include_fragment_spread() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    ...F @skip(if: true) @include(if: true)
                }
            }
            fragment F on User {
                age
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_and_include_inline_fragment() {
        let doc = parse_operation_document(
            "
            query {
                me {
                    id name type
                    ... @skip(if: true) @include(if: true) {
                        age
                    }
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_fragment_spread_variable() {
        let doc = parse_operation_document(
            "
            query($skip: Boolean!) {
                me {
                    id name type
                    ...F @skip(if: $skip)
                }
            }
            fragment F on User {
                age
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn include_fragment_spread_variable() {
        let doc = parse_operation_document(
            "
            query($include: Boolean!) {
                me {
                    id name type
                    ...F @include(if: $include)
                }
            }
            fragment F on User {
                age
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_inline_fragment_variable() {
        let doc = parse_operation_document(
            "
            query($skip: Boolean!) {
                me {
                    id name type
                    ... @skip(if: $skip) {
                        age
                    }
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn include_inline_fragment_variable() {
        let doc = parse_operation_document(
            "
            query($include: Boolean!) {
                me {
                    id name type
                    ... @include(if: $include) {
                        age
                    }
                }
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }

    #[test]
    fn skip_and_include_fragment_spread_variable() {
        let doc = parse_operation_document(
            "
            query($flag: Boolean!) {
                me {
                    id name
                    ...F @skip(if: $flag)
                    ... @include(if: $flag) {
                        age
                    }
                }
            }
            fragment F on User {
                type
            }
            ",
        )
        .unwrap();
        let printed = print_document_default(&doc);
        assert_snapshot!(printed);
    }
}

mod import_fragments {
    use super::*;

    #[test]
    fn import_fragment() {
        let doc = parse_operation_document(
            "
            #import UserProfile from \"./user-profile.graphql\"
            query myQuery {
                me {
                    id 
                    ...UserProfile
                }
            }
            ",
        )
        .unwrap();
        let (doc, extensions) = resolve_operation_extensions(doc).unwrap();
        let doc = (Path::new("/path/to/main.graphql"), &doc, &extensions);
        let resolved = resolve_operation_imports(doc, &TestOperationResolver).unwrap();
        let printed = print_document(&resolved, OperationTypePrinterOptions::default());
        assert_snapshot!(printed);
    }
}

struct TestOperationResolver;
impl<'src> OperationResolver<'src> for TestOperationResolver {
    fn resolve(
        &self,
        path: &Path,
    ) -> Option<(
        &OperationDocument<'src>,
        &nitrogql_semantics::OperationExtension<'src>,
    )> {
        match path.to_str().unwrap() {
            "/path/to/user-profile.graphql" => Some(static_parse(
                r#"
fragment UserProfile on User {
    name
    age
}
"#,
                1,
            )),
            _ => None,
        }
    }
}

fn static_parse(
    code: &'static str,
    file_id: usize,
) -> (
    &'static OperationDocument<'static>,
    &'static OperationExtension<'static>,
) {
    set_current_file_of_pos(file_id);
    let doc = parse_operation_document(code).unwrap();
    let (doc, extensions) = resolve_operation_extensions(doc).unwrap();
    let (doc, extensions) = (Box::leak(Box::new(doc)), Box::leak(Box::new(extensions)));
    (doc, extensions)
}

fn print_document_default(document: &OperationDocumentExt) -> String {
    set_current_file_of_pos(0);
    let (document, extensions) = resolve_operation_extensions(document.clone()).unwrap();
    let document = resolve_operation_imports(
        (Path::new("/path/to/main.graphql"), &document, &extensions),
        &TestOperationResolver,
    )
    .unwrap();
    print_document(&document, OperationTypePrinterOptions::default())
}

fn print_document(document: &OperationDocument, options: OperationTypePrinterOptions) -> String {
    let mut result = String::new();
    let mut writer = JustWriter::new(&mut result);
    let schema = type_system();
    let schema = ast_to_type_system(&schema);
    print_types_for_operation_document(options, &schema, document, &mut writer);
    result
}
