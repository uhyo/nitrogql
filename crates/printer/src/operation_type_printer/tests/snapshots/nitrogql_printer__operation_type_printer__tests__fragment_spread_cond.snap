---
source: crates/printer/src/operation_type_printer/tests/mod.rs
expression: printed
---
import type { TypedDocumentNode } from "@graphql-typed-document-node/core";
import type * as Schema from "";

type FooBar123Result = Schema.__SelectionSet<Schema.Query, {
  me: Schema.__SelectionSet<Schema.User, {
    id: Schema.ID;
    name: Schema.String;
    type: Schema.UserType;
    age: Schema.Int | null;
    posts: (Schema.__SelectionSet<Schema.User, {
      id: Schema.ID;
      id: Schema.ID;
    }, {}> | Schema.__SelectionSet<Schema.Bot, {
      id: Schema.ID;
      id: Schema.ID;
    }, {}> | Schema.__SelectionSet<Schema.Post, {
      id: Schema.ID;
      id: Schema.ID;
      title: Schema.String;
      body: Schema.String;
    }, {}>)[];
  }, {}>;
}, {}>;

type FooBar123Variables = {};

declare const FooBar123Query: TypedDocumentNode<FooBar123Result, FooBar123Variables>;

export { FooBar123Query as default };

export type F = Schema.__SelectionSet<Schema.User, {
  id: Schema.ID;
}, {}> | Schema.__SelectionSet<Schema.Bot, {
  id: Schema.ID;
}, {}> | Schema.__SelectionSet<Schema.Post, {
  id: Schema.ID;
}, {}>;

export const F: TypedDocumentNode<F, never>;

export type P = Schema.__SelectionSet<Schema.Post, {
  title: Schema.String;
  body: Schema.String;
}, {}>;

export const P: TypedDocumentNode<P, never>;


